src/Input/Select/Select.tsx

```diff
@@ -15,346 +15,287 @@ import {
     }
-  };
+  }, [value, defaultValue]);
 
-  handleFocusOut = (
-    onBlur: (e: React.FocusEvent<HTMLInputElement>) => void
-  ) => {
-    const listener = (e: React.FocusEvent<HTMLInputElement>) => {
-      this.setState({
-        floating: e.target.value.length > 0,
-      });
+  const handleFocusOut = React.useCallback(
+    (e: React.FocusEvent<HTMLInputElement>) => {
+      setFloating(e.target.value.length > 0);
 
       if (onBlur !== undefined) {
         return onBlur(e);
       }
-    };
-
-    return listener;
-  };
+    },
+    [onBlur]
+  );
 
-  handleFocus = (onFocus: (e: React.FocusEvent<HTMLInputElement>) => void) => {
-    const listener = (e: React.FocusEvent<HTMLInputElement>) => {
-      this.setState({
-        isFocus: true,
-      });
+  const handleFocus = React.useCallback(
+    (e: React.FocusEvent<HTMLInputElement>) => {
+      setIsFocus(true);
 
       if (onFocus !== undefined) {
         return onFocus(e);
       }
-    };
+    },
+    [onFocus]
+  );
 
-    return listener;
-  };
-
-  handleChange = (
-    onChange: (e: React.ChangeEvent<HTMLInputElement>) => void
-  ) => {
-    const listener = (e: React.ChangeEvent<HTMLInputElement>) => {
-      const { children } = this.props;
+  // Should be called when the user types into the input
+  const handleChange = React.useCallback(
+    (e: React.ChangeEvent<HTMLInputElement>) => {
       const childrenArray = React.Children.toArray(children) as Array<
         React.ReactElement<SelectItemProps>
       >;
 
-      this.setState({
-        selectedValue: e.target.value,
-        filterValue: childrenArray.filter(data =>
+      setSelectedValue(e.target.value);
+      setFilterValue(
+        childrenArray.filter(data =>
           data.props.children
             .toLowerCase()
             .includes(e.target.value.toLowerCase())
-        ),
-        cursor: 0,
-      });
+        )
+      );
+      setCursor(0);
 
       if (onChange !== undefined) {
-        return onChange(e);
+        return onChange(e.target.value);
       }
-    };
-
-    return listener;
-  };
-
-  handleClick = (
-    onOptionClick: (e: React.MouseEvent<HTMLLIElement, MouseEvent>) => void
-  ) => {
-    const listener = (e: React.MouseEvent<HTMLLIElement, MouseEvent>) => {
-      const { children, onChange } = this.props;
-      this.setState({
-        selectedValue: e.currentTarget.innerText,
-        filterValue: React.Children.map(children, data => data),
-        isFocus: false,
-      });
+    },
+    [children, onChange]
+  );
+
+  const getActiveElement = React.useCallback(() => {
+    return node.current.querySelector('.active') as HTMLLIElement;
+  }, []);
+
+  // Should be called when the user selects an option
+  const handleSelect = React.useCallback(
+    (e?: void) => {
+      const activeElement = getActiveElement();
+      const selectedValue = activeElement.innerText;
+      // Need  || '' here to ensure that the select input is never uncontrolled,
```

Comment:
Hmm, when will `selectedValue` be nil?

If it is defensive programming, I think we can just make `<SelectInput value={selectedValue || ''} // ... />` instead, then it might be safer.

Reply:
No this is some weird issue with React. Because you cannot change from uncontrolled to controlled component. `selectedValue` is undefined at some point before the user types, so the underlying select input would be interpreted as uncontrolled first. This simply makes sure that that never happens.

Reply:
Yeah, I knew the point for controlled component, but after I played around with `Select` I still cannot understand how `selectedValue` can be `undefined` 🤔

Reply:
It happens in the tests. There are seven test cases were I could find this issue. They are the ones who use either `setupSelectOptionFromMenuWithKeyHandling` or `setupSelectOptionFromMenu`. Not quite sure why this is.

Reply:
Okay I found out that the `innerText` is sometimes undefined. `textContent` works better here. That's what's used by the tests too.

Reply:
> Bug: Use the keyboard to select option, the floating label doesn't work well.
> Btw, would you mind adding a test case to catch this haha?
> ![image](https://user-images.githubusercontent.com/13249748/77400462-2d694200-6de6-11ea-980b-3d63987add72.png)

Fixed